src/Routing.purs

```diff
@@ -1,67 +1,92 @@
-module Routing (
-  hashChanged,
-  hashes,
-  matches,
-  matches',
-  match,
-  matchWith,
-  matchesAff,
-  matchesAff'
+module Routing
+  ( RoutingEffects
+  , foldHashes
+  , hashes
+  , matches
+  , matchesWith
+  , match
+  , matchWith
   ) where
 
-import Control.Monad.Aff (Aff, makeAff, nonCanceler)
+import Prelude
+
 import Control.Monad.Eff (Eff)
-import Data.Either (Either(..), either)
-import Data.Maybe (Maybe(..))
-import Data.String.Regex as R
-import Data.String.Regex.Flags as RF
-import Data.Tuple (Tuple(..))
-import Prelude (Unit, const, pure, unit, ($), (<$))
+import Control.Monad.Eff.Ref (REF, newRef, readRef, writeRef)
+import DOM (DOM)
+import DOM.Event.EventTarget (addEventListener, eventListener, removeEventListener)
+import DOM.HTML (window)
+import DOM.HTML.Event.EventTypes (hashchange)
+import DOM.HTML.Types (windowToEventTarget)
+import Data.Either (Either)
+import Data.Foldable (class Foldable, find)
+import Data.Maybe (Maybe(..), maybe)
+import Global (decodeURIComponent)
+import Routing.Hash (getHash)
 import Routing.Match (Match, runMatch)
 import Routing.Parser (parse)
 
+type RoutingEffects eff =
+  ( dom :: DOM
+  , ref :: REF
+  | eff
+  )
 
-foreign import decodeURIComponent :: String -> String
-
-foreign import hashChanged :: forall e. (String -> String -> Eff e Unit) -> Eff e Unit
-
-
-hashes :: forall e. (String -> String -> Eff e Unit) -> Eff e Unit
-hashes cb =
-  hashChanged $ \old new -> do
-    cb (dropHash old) (dropHash new)
-  where dropHash h =
-          case R.regex "^[^#]*#" RF.noFlags of
-            Right regX -> R.replace regX "" h
-            Left _     -> h
-
-
--- | Stream of hash changed, callback called when new hash can be matched
--- | First argument of callback is `Just a` when old hash can be matched
--- | and `Nothing` when it can't.
-matches :: forall e a. Match a -> (Maybe a -> a -> Eff e Unit) -> Eff e Unit
-matches = matches' decodeURIComponent
-
-matches' :: forall e a. (String -> String) ->
-            Match a -> (Maybe a -> a -> Eff e Unit) -> Eff e Unit
-matches' decoder routing cb = hashes $ \old new ->
-  let mr = matchWith decoder routing
-      fst = either (const Nothing) Just $ mr old
-  in either (const $ pure unit) (cb fst) $ mr new
+-- | Folds effectfully over hash changes given a callback and an initial hash.
+-- | The provided String is the hash portion of the `Location` with the '#'
+-- | prefix stripped. Returns an effect which will remove the listener.
+foldHashes
+  :: forall eff a
+   . (a -> String -> Eff (RoutingEffects eff) a)
+  -> (String -> Eff (RoutingEffects eff) a)
+  -> Eff (RoutingEffects eff) (Eff (RoutingEffects eff) Unit)
+foldHashes cb init = do
```

Comment:
Looks like this is adhoc `Program` from ps-coroutines, isn't it? 

I mean `Producer` that emits `String`s and consumer that calls callback.

Reply:
Maybe, but that's a significantly heavier API, and I don't see the advantage of using it when the only thing we are doing here is attaching a callback. Coroutines require `Aff` as well (for `Parallel`), and then there's no easy way to add finalizers to them. `purescript-aff-coroutines` just terminate and then leak since the `Eff` callbacks are still active,